src/core/assistant-message/presentAssistantMessage.ts

```diff
@@ -43,6 +43,105 @@ import { applyDiffTool as applyDiffToolClass } from "../tools/ApplyDiffTool"
 import { isNativeProtocol } from "@roo-code/types"
 import { resolveToolProtocol } from "../../utils/resolveToolProtocol"
 
+/**
+ * Finds the correct tool_use_id from the last assistant message in API history.
+ * This is crucial for ensuring tool_result blocks match existing tool_use blocks,
+ * especially when editing messages in completed tasks.
+ *
+ * When a user edits a message in a completed task:
+ * 1. API history is truncated (keeping assistant's tool_use with id=A)
```

Comment:
Why do we keep the assistant's tool use?

Reply:
Overall this just feels like a lot of special case logic. Is there a way that we could treat it more like a checkpoint rollback following by a new user message and not need any special casing aside from what's already in place to handle that case?